proposals/p3720.md

```diff
@@ -119,22 +120,33 @@ operators. There are three different interfaces used, depending on whether `x`
 is a value expression, a reference expression, or a facet:
 
 ```carbon
-// For a value expression `x` with type `T` and an expression `y` of type `U`,
-// `x.(y)` is `y.((U as ValueBind(T)).Op)(x)`
-interface ValueBind(T:! type) {
+// This determines the type of the result of binding. It is a
+// separate interface shared by `ValueBind` and `RefBind` to
+// ensure they produce the same result type. We don't want the
+// type of an expression to depend on the expression category
+// of the arguments.
+interface Bind(T:! type) {
   let Result:! type;
+}
+
+// For a value expression `x` with type `T` and an expression
+// `y` of type `U`, `x.(y)` is `y.((U as ValueBind(T)).Op)(x)`
+interface ValueBind(T:! type) {
+  extend Bind(T);
   fn Op[self: Self](x: T) -> Result;
 }
 
-// For a reference expression `x` using a binding `var x: T` and an expression `y` of type `U`,
-// `x.(y)` is `*y.((U as RefBind(T)).Op)(&x)`
+// For a reference expression `x` using a binding `var x: T`
+// and an expression `y` of type `U`, `x.(y)` is
+// `*y.((U as RefBind(T)).Op)(&x)`
 interface RefBind(T:! type) {
-  let Result:! type;
+  extend Bind(T);
   fn Op[self: Self](p: T*) -> Result*;
 }
 
-// For a facet value, which includes all type values, `T` and an expression `y` of type `U`,
-// `T.(y)` is `y.((U as TypeBind(T)).Op)()`.
+// For a facet value, which includes all type values, `T` and
+// an expression `y` of type `U`, `T.(y)` is
+// `y.((U as TypeBind(T)).Op)()`.
 interface TypeBind(T:! type) {
   let Result:! type;
```

Comment:
Are there coherence concerns here? For a given `x.(y)`, we might not know whether `x` is a facet value.

I don't think we can use `Bind` here, because we'd want to require that if `M impls TypeBind(T) with .Result = R` then `M impls Bind(type) with .Result = R` (and in general, `forall [F is a facet type] M impls Bind(F) with .Result = R`). But `R` needs to depend on `T`, and `T` isn't in scope in that impl of `Bind`.

I suppose the upshot of this is that the "facet on the left of `.`" check must be done in the definition of the generic, and if it turns out that the left-hand expression is a facet once we know its type, that's just too late, and you get the `ValueBind` / `RefBind` behavior instead. How significant of an issue is that? Something similar comes up elsewhere, where we say a direct member access into a facet doesn't use an overloadable binding operator.

Reply:
My lack of worry mostly comes from not having an alternative to compare it to.

Reply:
Added a sentence to the note about `BindToType`.